src/System.IO.Abstractions.Extensions/IDirectoryInfoExtensions.cs

```diff
@@ -199,18 +199,36 @@ public static string[] DiffPaths(this IDirectoryInfo ancestor, IDirectoryInfo ch
         /// </summary>
         /// <param name="source">Source directory</param>
         /// <param name="destination">Destination directory</param>
-        /// <param name="recursive">If true the copy will be recursive and will include subfolders of <paramref name="info"/>. Defaults to false</param>
+        /// <param name="recursive">If true the copy will be recursive and will include subfolders of <paramref name="source"/>. Defaults to false</param>
+        /// <param name="overwrite">If true the copy will overwrite any existing files in <paramref name="destination"/>. Defaults to false</param>
         /// <param name="filesSearchPattern">Search pattern to apply when searching files, defaults to '*'</param>
         /// <param name="directoriesSearchPattern">Search pattern to apply when searching directories, defaults to '*'</param>
         public static void CopyTo(
             this IDirectoryInfo source,
             IDirectoryInfo destination,
             bool recursive = false,
+            bool overwrite = false,
             string filesSearchPattern = "*",
             string directoriesSearchPattern = "*")
         {
+            if (!overwrite)
```

Comment:
I don't understand the need for this whole new section  as the CopyTo call already has the overwrite parameter and will throw an exception

Reply:
This adds the behaviour of not doing a partial copy of a directory if some files already exist whilst overwrite is not set. Currently, the loop will fail on the first file it is unable to copy.

Take a source directory containing `a`, b` and `c`, and a destination directory containing `b`. Existing behaviour would copy `a` then fail on `b` as it cannot be overwritten, leaving the destination containing `a` and the old `b`. This addition loop makes it so that the destination directory would contain just the old `b` after the operation failed since file existence is checked prior to the copy.

Reply:
Yes, it would increase the complexity during the default runtime. Let's remove it for now and address it in a separate PR.